extensions/amp-web-push/0.1/amp-web-push.service-worker.js

```diff
@@ -0,0 +1,155 @@
+  which broadcasts the reply back to the AMP page.
+ */
+
+const WorkerMessengerCommand = {
+  /*
+    Used to request the current subscription state.
+   */
+  AMP_SUBSCRIPION_STATE: 'amp-web-push-subscription-state',
+  /*
+    Used to request the service worker to subscribe the user to push.
+    Notification permissions are already granted at this point.
+   */
+  AMP_SUBSCRIBE: 'amp-web-push-subscribe',
+  /*
+    Used to unsusbcribe the user from push.
+   */
+  AMP_UNSUBSCRIBE: 'amp-web-push-unsubscribe',
+};
+
+/*
+  According to
+  https://w3c.github.io/ServiceWorker/#run-service-worker-algorithm:
+
+  "user agents are encouraged to show a warning that the event listeners
+  must be added on the very first evaluation of the worker script."
+
+  We have to register our event handler statically (not within an
+  asynchronous method) so that the browser can optimize not waking up the
+  service worker for events that aren't known for sure to be listened for.
+
+  Also see: https://github.com/w3c/ServiceWorker/issues/1156
+*/
+self.addEventListener('message', event => {
+  /*
+    Messages sent from amp-web-push have the format:
+
+    - command: A string describing the message topic (e.g.
+      'amp-web-push-subscribe')
+
+    - payload: An optional JavaScript object containing extra data relevant to
+      the command.
+   */
+  const {command, payload} = event.data;
+
+  switch (command) {
+    case WorkerMessengerCommand.AMP_SUBSCRIPION_STATE:
+      onMessageReceivedSubscriptionState(payload);
+      break;
+    case WorkerMessengerCommand.AMP_SUBSCRIBE:
+      onMessageReceivedSubscribe(payload);
+      break;
+    case WorkerMessengerCommand.AMP_UNSUBSCRIBE:
+      onMessageReceivedUnsubscribe(payload);
+      break;
+  }
+});
+
+/**
+  Broadcasts a single boolean describing whether the user is subscribed.
+ */
+function onMessageReceivedSubscriptionState() {
+  let retrievedPushSubscription = null;
+  self.registration.pushManager.getSubscription()
+      .then(pushSubscription => {
+        retrievedPushSubscription = pushSubscription;
+        if (!pushSubscription) {
+          return null;
+        } else {
+          return self.registration.pushManager.permissionState(
+              pushSubscription.options
+        );
+        }
+      }).then(permissionStateOrNull => {
+        if (permissionStateOrNull == null) {
+          broadcastReply(WorkerMessengerCommand.AMP_SUBSCRIPION_STATE, false);
+        } else {
+          const isSubscribed = !!retrievedPushSubscription &&
+            permissionStateOrNull === 'granted';
+          broadcastReply(WorkerMessengerCommand.AMP_SUBSCRIPION_STATE,
+              isSubscribed);
+        }
+      });
+}
+
+/**
+  Subscribes the visitor to push.
+
+  The broadcast value is null (not used in the AMP page).
+ */
+function onMessageReceivedSubscribe() {
+  self.registration.pushManager.subscribe({
+    userVisibleOnly: true,
+    applicationServerKey: 'fake-demo-key',
```

Comment:
Added a detailed comment recommending using existing subscription code for users that have it, and otherwise linking to guides on how to obtain an `applicationServerKey`. Ditto for the `unsubscribe()` method later down.